### PR TITLE
Implementing parser changes for "ref readonly"

### PR DESCRIPTION
**Pull request by OmarTawfik**

### CODE REVIEW

**Comment by OmarTawfik** [comment]
@dotnet/roslyn-compiler

File: src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ public virtual ImmutableArray<AttributeData> GetReturnTypeAttributes()
         public abstract int Arity { get; }
         public abstract bool ReturnsVoid { get; }
         public abstract bool ReturnsByRef { get; }
-        public abstract bool ReturnsByReadonlyRef { get; }
+        public abstract bool ReturnsByRefReadonly { get; }
```
Comment (jaredpar) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs]: Do we need an API that returns true if it returns `ref` or `ref readonly`?

Comment (VSadov) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs]: Yes, my change changes this API. ReturnsByRef would return `true` regardless if `readonly` or not, and then there will be another method to check for `readonly`

Usage shows it is more convenient. But that yet may change. 
That is why we are not asking yet for a formal review of this API.


---
In reply to: [102998261](https://github.com/dotnet/roslyn/pull/17337#discussion_r102998261) [](ancestors = 102998261)

Comment (VSadov) [inline src/Workspaces/Core/Portable/CodeGeneration/Symbols/CodeGenerationAbstractMethodSymbol.cs]: This change is ok for now

---
In reply to: [103024354](https://github.com/dotnet/roslyn/pull/17337#discussion_r103024354) [](ancestors = 103024354,102998261)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 2]
```diff
@@ -8799,8 +8814,8 @@ private void ParseDeclarationModifiers(SyntaxListBuilder list)
             SyntaxKind k;
             while (IsDeclarationModifier(k = this.CurrentToken.ContextualKind) || IsAdditionalLocalFunctionModifier(k))
             {
-                // "readonly ref" is not a modifier, in our syntax.
-                if (k == SyntaxKind.ReadOnlyKeyword && this.PeekToken(1).Kind == SyntaxKind.RefKeyword)
+                // "ref readonly" is not a modifier, in our syntax.
+                if (k == SyntaxKind.RefKeyword && this.PeekToken(1).Kind == SyntaxKind.ReadOnlyKeyword)
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I think this check is no longer needed. `ref` is not a declaration modifier.

**Comment by VSadov** [comment]
I think we should merge this so that we could rebase two other pending changes on top of this.
Any feedback could be handled in a subsequent change.

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 2 above)
Comment (OmarTawfik) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Per discussion, @VSadov will handle that change.

---
In reply to: [103036678](https://github.com/dotnet/roslyn/pull/17337#discussion_r103036678) [](ancestors = 103036678)

**Pull request merged**